Update package guidelines for .NET Core 3.0 (#13196)
dotnet/docs: docs/core/build/distribution-packaging.md

> ├── sdk

Reviewer: preivew 8 has added a `templates` directory next to the sdk: https://github.com/dotnet/core/issues/3184#issuecomment-521702831

Author: @dagood, from our source-build it seems templates are under a runtime versioned folder:
```
$ ls /usr/lib64/dotnet/templates/
3.0.0-preview8-013656
```
Why is it like this, and not the sdk version?

What package will the templates be part of? `dotnet-sdk-[major].[minor]` or their own package (`dotnet-templates-[major].[minor]`?).

Reviewer 2: > What package will the templates be part of? `dotnet-sdk-[major].[minor]` or their own package (`dotnet-templates-[major].[minor]`?).

Peeking at an SDK Deb package from the link table at https://github.com/dotnet/core-sdk, it's in the SDK package for 3.0 preview 9.

It would make sense to me, though, to put them in a different package so users can choose different major-minor versions of templates without installing multiple SDKs. I *think* this is a general direction we want to take for the Microsoft build too, @leecow correct me if I'm wrong.

Author: > @tmds I asked the same question about the funny version number here: dotnet/cli#12275

It doesn't make sense for me that there is a runtime version there because it's content that you may want to change with an sdk version, without there being a the runtime version change.

Anyway, the general rule is: if something is under a folder named `[major].[minor].[patch]-[suffix]` it ends up in a package named `packagename-[major].[minor]`.

If we can use the same version for all the packages (cfr https://github.com/dotnet/docs/pull/13196#discussion_r317047143), the package maintainer doesn't need to worry about propagating the `[patch]-[suffix]` into the package versions.

Author: I have added a `dotnet-templates-[major].[minor]` package.

- (3) **sdk/\<sdk version>** The SDK (also known as "the tooling") is a set of managed tools that are used to write and build .NET Core libraries and applications. The SDK includes the .NET Core Command-line interface (CLI), the managed languages compilers, MSBuild, and associated build tasks and targets, NuGet, new project templates, and so on.

> - (4) **sdk/NuGetFallbackFolder** contains a cache of NuGet packages used by an SDK during the restore operation, such as when running `dotnet restore` or `dotnet build /t:Restore`. This folder is no longer used with .NET Core 3.0+. For earlier versions, it could not be built from source.

Author: [source-build](https://github.com/dotnet/source-build/) isn't able to build the content of this folder.

> - (4) **sdk/NuGetFallbackFolder** contains a cache of NuGet packages used by an SDK during the restore operation, such as when running `dotnet restore` or `dotnet build /t:Restore`. This folder is no longer used with .NET Core 3.0+. For earlier versions, it could not be built from source.

Reviewer 2: Maybe invert to make it a bit simpler, and add a little bit about why it can't be built from source (which also clarifies the meaning if the reader doesn't have context)?

```suggestion
- (4) **sdk/NuGetFallbackFolder** contains a cache of NuGet packages used by an SDK during the restore operation, such as when running `dotnet restore` or `dotnet build /t:Restore`. This folder is only used prior to .NET Core 3.0. It can't be built from source, because it contains prebuilt binary assets from `nuget.org`.
```

- (4) **sdk/NuGetFallbackFolder** contains a cache of NuGet packages used by an SDK during the restore operation, such as when running `dotnet restore` or `dotnet build /t:Restore`. This folder is only used prior to .NET Core 3.0. It can't be built from source, because it contains prebuilt binary assets from `nuget.org`.

The **shared** folder contains frameworks. A shared framework provides a set of libraries at a central location so they can be used by different applications.

- (5) **shared/Microsoft.NETCore.App/\<runtime version>** This framework contains the .NET Core runtime and supporting managed libraries.

- (6) **shared/Microsoft.AspNetCore.{App,All}/\<aspnetcore version>** contains the ASP.NET Core libraries. The libraries under `Microsoft.AspNetCore.App` are developed and supported as part of the .NET Core project. The libraries under `Microsoft.AspNetCore.All` are a superset that also contains third-party libraries.

> - (7) **shared/Microsoft.Desktop.App/\<desktop app version>** contains the Windows desktop libraries. This isn't included on non-Windows platforms.

Reviewer 3: Should it be noted that some directories/components are available at .NET Core 3.0+?

Reviewer 3: nevermind ... the title of the doc is 3.0

| dotnet-sdk-[major].[minor]        | dotnet-sdk-2.1     | Latest sdk for specific runtime | (3),(4),(11),(12)  | aspnet-runtime-[major].[minor], dotnet-netstd-pack-[netstd_major].[netstd_minor] | \<sdk version>     |
| aspnet-runtime-[major].[minor]    | aspnet-runtime-2.1 | Specific ASP.NET Core runtime   | (6),[(7)]          | dotnet-runtime-[major].[minor]                 | \<runtime version> |
| dotnet-runtime-[major].[minor]    | dotnet-runtime-2.1 | Specific runtime                | (5)                | host-fxr:\<runtime version>+                   | \<runtime version> |
| dotnet-host-fxr                   | dotnet-host-fxr    | _dependency_                    | (2)                | host:\<runtime version>+                       | \<runtime version> |
| dotnet-host                       | dotnet-host        | _dependency_                    | (1),(8),(9),(10)   |                                                | \<runtime version> |
| dotnet-desktop-pack-[major.minor] |                    | _dependency_                    | (12)               |                                                | \<desktop version>  |

> | dotnet-netstd-pack-[major.minor]  |                    | _dependency_                    | (13)               |                                                | \<netstd version>  |

Reviewer 2: These are called `netstandard-targeting-pack` and `windowsdesktop-targeting-pack` in Microsoft builds. I can see how a prefix would be a good idea for discoverability (especially when on a "real" package repository) but I'm not a fan of the abbreviations here.

`netstd`: I haven't seen this before, I searched and found that we don't want to use it: https://github.com/dotnet/corefx/pull/4362#discussion-diff-44104229.

`pack`: there are multiple kinds of packs, only including `pack` in the package name could be confusing. In particular, someone installing this might expect runtime packs.

`desktop`: well, we don't currently plan to make this targeting pack available on Linux because there are some build-time Windows dependencies: https://github.com/dotnet/core-setup/pull/6198. If it somehow became a critical scenario to allow someone to target WindowsDesktop when building on Linux, though, including `windows` seems like a good idea. It would match the name of the framework, and avoids being misinterpreted into a false expectation that it will work on non-Windows platforms.

Author: Thanks @dagood. I've added separate packages for aspnet and netcoreapp, and changed the naming a bit.

> desktop: well, we don't currently plan to make this targeting pack available on Linux because there are some build-time Windows dependencies:

We need to look into this. The 3.0 preview 6 build @omajid did for Fedora 30 includes the desktop targeting pack.

Reviewer: > desktop: well, we don't currently plan to make this targeting pack available on Linux because there are some build-time Windows dependencies:

@dleeapho @crummel Any ideas about this? Should source-build be building `Microsoft.WindowsDesktop.App.Ref` on Linux?

Author: @dagood do you know why we have the windows targeting pack in our source-build preview6? Does it need a fix somewhere?

Reviewer 2: The intent is definitely not to build it: https://github.com/dotnet/core-setup/blob/master/src/pkg/projects/dir.props#L32-L35. But I'm also wondering, where are you seeing it, what type of artifact?

This is the new nupkgs that source-build infra detects, at least--no windowsdesktop:

```
  [23:04:07.40] Building 'core-setup' to produce tarball...done
  New NuGet package(s) after building core-setup:
    -> Microsoft.DotNet.PlatformAbstractions 3.0.0-preview6-27804-01
    -> Microsoft.Extensions.DependencyModel 3.0.0-preview6-27804-01
    -> Microsoft.NET.HostModel 3.0.0-preview6-27804-01
    -> Microsoft.NETCore.App 3.0.0-preview6-27804-01
    -> Microsoft.NETCore.App.Host.centos.7-x64 3.0.0-preview6-27804-01
    -> Microsoft.NETCore.App.Internal 3.0.0-preview6-27804-01
    -> Microsoft.NETCore.App.Ref 3.0.0-preview6-27804-01
    -> Microsoft.NETCore.DotNetAppHost 3.0.0-preview6-27804-01
    -> Microsoft.NETCore.DotNetHost 3.0.0-preview6-27804-01
    -> Microsoft.NETCore.DotNetHostPolicy 3.0.0-preview6-27804-01
    -> Microsoft.NETCore.DotNetHostResolver 3.0.0-preview6-27804-01
    -> NETStandard.Library.Ref 2.1.0-preview6-27804-01
    -> runtime.centos.7-x64.Microsoft.NETCore.App 3.0.0-preview6-27804-01
    -> runtime.centos.7-x64.Microsoft.NETCore.DotNetAppHost 3.0.0-preview6-27804-01
    -> runtime.centos.7-x64.Microsoft.NETCore.DotNetHost 3.0.0-preview6-27804-01
    -> runtime.centos.7-x64.Microsoft.NETCore.DotNetHostPolicy 3.0.0-preview6-27804-01
    -> runtime.centos.7-x64.Microsoft.NETCore.DotNetHostResolver 3.0.0-preview6-27804-01
```

Searching the logs, I see they're created as reference packages, but those are only meant for source-build bits to compile against:

```
  Created '/tb/tarball_246406/reference-packages/packages/Microsoft.WindowsDesktop.App.Ref.3.0.0-preview6-27804-01.nupkg'
  Created '/tb/tarball_246406/reference-packages/packages/Microsoft.WindowsDesktop.App.3.0.0-preview6-27804-01.nupkg'
```

(I'm a little surprised to see those there, though. The annotated prebuilt doesn't even say where they're used. Filed https://github.com/dotnet/source-build/issues/1128 to figure that out.)

Author: Our source-build 3.0 preview6 has a Microsoft.WindowsDesktop.App.Ref folder:
```
$ ls /usr/lib64/dotnet/packs
Microsoft.AspNetCore.App.Ref              Microsoft.NETCore.App.Ref         NETStandard.Library.Ref
Microsoft.NETCore.App.Host.fedora.30-x64  Microsoft.WindowsDesktop.App.Ref
```

Reviewer 2: I see... Core-SDK is downloading the WindowsDesktop packages in the production build, which is a bug. Filed https://github.com/dotnet/core-sdk/issues/2987.

Reviewer 4: @omajid Shipping the downloaded version is for sure a bug, I think we probably don't have to build it at all since we shouldn't be building anything Windows-specific.

| dotnet-sdk-[major]                | dotnet-sdk-2       | Latest sdk for runtime major    |                  | dotnet-sdk-[major].[latestminor]  | \<sdk version>     |
| dotnet-sdk-[major].[minor]        | dotnet-sdk-2.1     | Latest sdk for specific runtime | (3),(4),(13)     | aspnet-runtime-[major].[minor], dotnet-netcoreapp-targeting-pack[major].[minor], dotnet-aspnet-targeting-pack[major].[minor], dotnet-netstandard-targeting-pack-[netst_major].[netstd_minor] | \<sdk version>     |
| aspnet-runtime-[major].[minor]    | aspnet-runtime-2.1 | Specific ASP.NET Core runtime   | (6),[(7)]        | dotnet-runtime-[major].[minor]    | \<runtime version> |
| dotnet-runtime-[major].[minor]    | dotnet-runtime-2.1 | Specific runtime                | (5)              | host-fxr:\<runtime version>+      | \<runtime version> |
| dotnet-host-fxr                   | dotnet-host-fxr    | _dependency_                    | (2)              | host:\<runtime version>+          | \<runtime version> |
| dotnet-host                       | dotnet-host        | _dependency_                    | (1),(8),(9),(10) |                                   | \<runtime version> |

> | dotnet-netcoreapp-targeting-pack-[major].[minor]    |  | _dependency_                    | (12)             |                                   | \<pack version>    |

Reviewer 2: This doesn't have `netcoreapp` in the Microsoft build:

```suggestion
| dotnet-targeting-pack-[major].[minor]    |  | _dependency_                    | (12)             |                                   | \<pack version>    |
```

I don't feel strongly either way--I think what matters here (as far as naming matters :smile:) is making these consistent, which you could do any number of ways:

```
dotnet-runtime-[major].[minor]
dotnet-targeting-pack-[major].[minor]
aspnet-runtime-[major].[minor]
aspnet-targeting-pack-[major].[minor]
```

Author: I'm not opinionated about the names either, I care about having the same names for Red Hat/Fedora/Microsoft packages.

Reviewer 2: Here are the Preview 6 RPMs I know about from Microsoft build output, for reference:

**ASP.NET Core (AspNetCore)**

| Name | Version | Release | Requires |
| --- | --- | --- | ---
| **aspnetcore-runtime-3.0** | 3.0.0 | 0.1.preview6.19307.2 | dotnet-runtime-3.0 |
| aspnetcore-targeting-pack-3.0 | 3.0.0 | 0.1.preview6.19307.2 | dotnet-targeting-pack-3.0 |

**SDK (Core-SDK)**

| Name | Version | Release | Requires |
| --- | --- | --- | ---
| **dotnet-sdk-3.0** | 3.0.100_preview6_012264 | 1 | dotnet-runtime-3.0, dotnet-targeting-pack-3.0, dotnet-apphost-pack-3.0, netstandard-targeting-pack-2.1, aspnetcore-runtime-3.0, aspnetcore-targeting-pack-3.0 |

**Runtime (Core-Setup)**

| Name | Version | Release | Requires |
| --- | --- | --- | ---
| dotnet-apphost-pack-3.0 | 3.0.0 | 0.1._preview6_27804_01 |  |
| dotnet-host | 3.0.0 | 0.1.preview6_27804_01 |  |
| dotnet-hostfxr-3.0 | 3.0.0 | 0.1.preview6_27804_01 | dotnet-host |
| **dotnet-runtime-3.0** | 3.0.0 | 0.1.preview6_27804_01 | dotnet-hostfxr-3.0, dotnet-runtime-deps-3.0 |
| dotnet-runtime-deps-3.0 | 3.0.0 | 0.1.preview6_27804_01 | `<per-distro>` |
| dotnet-targeting-pack-3.0 | 3.0.0 | 0.1._preview6_27804_01 |  |
| netstandard-targeting-pack-2.1 | 2.1.0 | 0.1._preview6_27804_01 |  |

```sh
rpm -qp --queryformat '| %{NAME} | %{VERSION} | %{RELEASE} | [%{REQUIRES}, ] |\n' *.rpm
# (With some cleanup)
```

---

The versions look odd (probably because we haven't put preview packages on a package repository so they aren't ironed out) but I think the granularity and dependencies are what we're expecting to go with. I bolded the expected "entry point" packages for install commands.

(Going into this, I assumed the existing set was common knowledge and this PR was about making changes for Red Hat and other distros, but I don't actually know how I would have generated those tables without access to internal data. 😕 We need [public BAR access](https://github.com/dotnet/arcade/issues/1818).)

Author: > I assumed the existing set was common knowledge and this PR was about making changes for Red Hat and other distros, but I don't actually know how I would have generated those tables without access to internal data.

There has been no public info about this, that's why I've started the PR.

Based on your list, I'll make these changes:
- `aspnetcore` vs `aspnet` -> `aspnetcore`
- remove `dotnet-` prefix
- `dotnet-hostfxr` vs `dotnet-host-fxr`, Microsoft is using the former, this doc (and Fedora/RHEL) the latter. -> changing to `dotnet-hostfxr`.

I think for Fedora/RHEL we'll stick to `dotnet-host-fxr` since we are using that name already.

- `dotnet-apphost-pack-3.0`, I'll add this.

The packs have _full version number folders:
```
├── Microsoft.AspNetCore.App.Ref
│   └── 3.0.0-preview6.19307.2
├── Microsoft.NETCore.App.Host.fedora.30-x64
│   └── 3.0.0-preview6-27804-01
├── Microsoft.NETCore.App.Ref
│   └── 3.0.0-preview6-27804-01
├── Microsoft.WindowsDesktop.App.Ref
│   └── 3.0.0-preview6-27804-01
└── NETStandard.Library.Ref
    └── 2.1.0-preview6-27804-01
```
Does that imply that the sdk supports some sort of _roll forward_ if successive versions of the same package use different version numbers here?
Or will this in be: `[major].[minor].0` for released versions, and not change for successive versions?

Reviewer 2: > The packs have _full version number folders: [...]
>
> Does that imply that the sdk supports some sort of _roll forward_ if successive versions of the same package use different version numbers here?
> Or will this in be: `[major].[minor].0` for released versions, and not change for successive versions?

https://github.com/dotnet/designs/pull/50 touches on the philosophy around this:

> Targeting pack package versions will generally not increase past major.minor.0 where major.minor matches the the corresponding two-part TFM. For example, when there is a 3.0.1 .NET Core runtime, the targeting pack will likely remain at 3.0.0. This follows from the fact that targeting packs represent public API surface, which must not change in a patch version of the runtime. (With that said, we can reserve the right to modify the targeting pack in a patch release to fix a severe bug. In the rare event that this occurs, the targeting pack patch version could be incremented past 0.)
> 
> During the prerelease phase of a new major.minor version of the runtime, the targeting pack will version 1:1 with the runtime. This is necessary as different prerelease versions will have different surface area.

The SDK has the version in `Microsoft.NETCoreSdk.BundledVersions.props` as of preview6 ([generated here](https://github.com/dotnet/core-sdk/blob/master/src/redist/targets/GenerateBundledVersions.targets)):

```xml
    <KnownFrameworkReference Include="Microsoft.NETCore.App"
                              TargetFramework="netcoreapp3.0"
                              RuntimeFrameworkName="Microsoft.NETCore.App"
                              DefaultRuntimeFrameworkVersion="3.0.0-preview6-27804-01"
                              LatestRuntimeFrameworkVersion="3.0.0-preview6-27804-01"
                              TargetingPackName="Microsoft.NETCore.App.Ref"
                              TargetingPackVersion="3.0.0-preview6-27804-01"
                              AppHostPackNamePattern="runtime.**RID**.Microsoft.NETCore.DotNetAppHost"
                              AppHostRuntimeIdentifiers="linux-arm;linux-arm64;linux-musl-arm64;linux-musl-x64;linux-x64;osx-x64;rhel.6-x64;tizen.4.0.0-armel;tizen.5.0.0-armel;win-arm;win-arm64;win-x64;win-x86"
                              RuntimePackNamePatterns="runtime.**RID**.Microsoft.NETCore.App;runtime.**RID**.Microsoft.NETCore.DotNetHostResolver;runtime.**RID**.Microsoft.NETCore.DotNetHostPolicy"
                              RuntimePackRuntimeIdentifiers="linux-arm;linux-arm64;linux-musl-arm64;linux-musl-x64;linux-x64;osx-x64;rhel.6-x64;tizen.4.0.0-armel;tizen.5.0.0-armel;win-arm;win-arm64;win-x64;win-x86"
                              PackagesToReference="Microsoft.NETCore.App/3.0.0-preview6-27804-01"
                              IsTrimmable="true"
                              />
```

I imagine there's a project-driven way to use a specific version, but I don't know of any roll forward based on what's on disk.

Author: > For example, when there is a 3.0.1 .NET Core runtime, the targeting pack will likely remain at 3.0.0. This follows from the fact that targeting packs represent public API surface, which must not change in a patch version of the runtime.

ok, this matches my expectations.

This was about the version that shows up in the _folders_. Now for the version we put onto the pack _packages_. I think for practical reasons we could use the `<sdk version>`? It seems simpler if successive versions of .NET Core can build a package with a higher version (even if the content is the same)? Or should we for some reason version these based on what is in the package?

Reviewer 2: > successive versions of .NET Core can build a package with a higher version (even if the content is the same)?

This part sounds very similar to the CoreFX platform extension package problem: https://github.com/dotnet/source-build/issues/923. The core is that there's no assurance it's possible to build old assets using a new commit because the Microsoft build process doesn't exercise it.

Independent of that, IMO a reason to not use the SDK version is to let users get lower-versioned targeting packs without having to look up too much about how versioning works. I.e. if they want to target netstandard 2.1, they get `netstandard-targeting-pack-2.1`. They don't have to figure out what SDK version was built with that targeting pack (3.0).

What makes it more practical to use SDK version? Would it help to split up source-build into "runtime" (includes runtime targeting packs), "netstandard", and "sdk" builds? (Or imagine it split up that way?)

Author: > the CoreFX platform extension package problem

I don't understand what this is about. How does source-build relate to _specific non-shared-framework packages_?

> IMO a reason to not use the SDK version is to let users get lower-versioned targeting packs without having to look up too much about how versioning works. 

Suppose there is no .NET Core on my machine, and I'd like to install .NET Core 3.1 sdk and 3.0 targeting pack. Then I'd:
```
$ dnf install dotnet-sdk-3.1 dotnet-targeting-pack-3.0
```
The versions the user cares about are part of the package names, the package versions don't show up here.

> What makes it more practical to use SDK version? Would it help to split up source-build into "runtime" (includes runtime targeting packs), "netstandard", and "sdk" builds? (Or imagine it split up that way?)

In once case source-build `3.0.100` and `3.0.200` both build a `dotnet-targeting-pack-3.0` with version `3.0.0` And in the other, the versions are `3.0.100` and `3.0.200`.
I was thinking that for a package maintainer, maybe it is simpler to not have to deal with packages that end up with the same version numbers.
But it seems this is already happening for the sdk/runtime (runtime `2.2.6` is used with sdk `2.2.108`, `2.2.205`, and `2.2.301`).

Reviewer 2: > > the CoreFX platform extension package problem
> 
> I don't understand what this is about. How does source-build relate to _specific non-shared-framework packages_?

I'm not quite sure what you mean by shared. But in any case, my point here was more about the problem of building runtime 3.0.1 (say there were no targeting pack fixes), but trying to build targeting packs that match the 3.0.0 outputs. This is what wouldn't reliably work for the same reasons as https://github.com/dotnet/source-build/issues/923. It seemed like some of this might depend on that working, so I wanted to bring it up, maybe I was mistaken. Either way, that's a big topic that calls for a general solution and doesn't need to be fleshed out here.

> The versions the user cares about are part of the package names, the package versions don't show up here.

Good clarification, I've always assumed the major/minor in the package name and the package version should match as some sort of convention, but I don't know a reason for it. I do agree that keeping the package version consistent across a build of the packages makes sense.

> In once case source-build `3.0.100` and `3.0.200` both build a `dotnet-targeting-pack-3.0` with version `3.0.0` And in the other, the versions are `3.0.100` and `3.0.200`.
> I was thinking that for a package maintainer, maybe it is simpler to not have to deal with packages that end up with the same version numbers.
> But it seems this is already happening for the sdk/runtime (runtime `2.2.6` is used with sdk `2.2.108`, `2.2.205`, and `2.2.301`).

Thanks, I agree. It's somewhat confusing that two builds produce the same packages, but at least with different package versions there's an obvious way to resolve it.

I think splitting up the builds resolves this too. But now I'm thinking it makes sense to document what can be done right now (one package version per build, based on the "root" SDK artifact) and come back to revise if splitting becomes possible.

Author: > I think splitting up the builds resolves this too. But now I'm thinking it makes sense to document what can be done right now (one package version per build, based on the "root" SDK artifact) and come back to revise if splitting becomes possible.

Yes, all packages could use the same version (incl runtime, targeting packs), this would be the tagged version of source-build, which is the sdk version.

The main value of different versions(/splitting) is that users don't have to download some of the packages over in case of an upgrade (because the content hasn't changed). I think the users don't care much about this.

Author: @leecow @dagood shall we use the same version for all packages (that is: sdk version)?

Author: Having different package versions for components that are built together adds complexity for the package maintainer.

For example, `netstandard-targeting-pack-2.1` for `3.0.100` sdk will have the version `2.1.0` in the filesystem layout. Probably the `3.0.101` sdk will also have `2.1.0` in the filesystem.

If we're using the `2.1.0` as the package version, what should the maintainer do when `3.0.101` builds a package with the same version?

If we're using the sdk version, because there is a higher version number from the sdk (`3.0.100` to `3.0.101`), he can just push a new package.

Author: > Note that the Microsoft build does separate the Runtime vs. SDK build

@dagood, I guess `Microsoft.NETCore.App.Ref`, `Microsoft.NETCore.App.Host.<rid>` are built with the .NET Core runtime?
`Microsoft.AspNetCore.App.Ref` is built with ASP.NET Core shared framework?
`templates` are built with the SDK?
What builds `NETStandard.Library.Ref`, `Microsoft.WindowsDesktop.App.Ref`?

Author: I've pushed a commit that sets the package version for:
`dotnet-apphost-pack-[major].[minor]`, `dotnet-targeting-pack-[major].[minor]` to \<runtime version>.
`aspnetcore-targeting-pack-[major].[minor]` to \<aspnet core runtime version>.
`netstandard-targeting-pack-[major].[minor]` and `dotnet-templates-[major].[minor]` to \<sdk version>.

| dotnet-aspnet-targeting-pack-[major].[minor]        |  | _dependency_                    | (11)             |                                   | \<pack version>    |
| dotnet-netstandard-targeting-pack-[netst_major].[netst_minor] || _dependency_            | (15)             |                                   | \<pack version>    |

Most distributions require all artifacts to be built from source. This has some impact on the packages:

- The third-party libraries under `shared/Microsoft.AspNetCore.All` can't be easily built from source. So that folder is omitted from the `aspnet-runtime` package.

- The `NuGetFallbackFolder` is populated using binary artifacts from `nuget.org`. It should remain empty.

Multiple `dotnet-sdk` packages may provide the same files for the `NuGetFallbackFolder`. To avoid issues with the package manager, these files should be identical (checksum, modification date, and so on).

### Preview versions

> Package maintainers may decide to provide preview versions of the shared framework and SDK. Preview releases may be provided using the `dotnet-sdk-[major].[minor]`, `aspnet-runtime-[major].[minor]`, or `dotnet-runtime-[major].[minor]` packages. For preview releases, the package version major must be set to zero. This way, the final release is installed as an upgrade of the package.

Reviewer: > For preview releases, the package version major must be set to zero

Is this something that's important to specify and include in this doc? Most distributions have their own policies on how non-final versions and upgrades are handled. This note either complements those distribution policies (not adding anything new) or it contradicts those policies (in which case the distributions will ignore this).

Fedora, for example, will use the release field, not version, for this. And they start with a zero prefix, but not zero.

Author: I wouldn't mind leaving out the whole preview release section. Aligning on the released packages is the primary target.

| aspnetcore-runtime-[major].[minor] | aspnetcore-runtime-2.1 | Specific ASP.NET Core runtime | (6),[(7)]      | dotnet-runtime-[major].[minor]     | \<runtime version> |
| dotnet-runtime-[major].[minor]     | dotnet-runtime-2.1 | Specific runtime                | (5)              | dotnet-hostfxr:\<runtime version>+ | \<runtime version> |
| dotnet-hostfxr                     | dotnet-hostfxr     | _dependency_                    | (2)              | host:\<runtime version>+           | \<runtime version> |
| dotnet-host                        | dotnet-host        | _dependency_                    | (1),(8),(9),(10) |                                    | \<runtime version> |
| dotnet-apphost-pack-[major].[minor]|                    | _dependency_                    | (13)             |                                    | \<pack version>    |
| dotnet-targeting-pack-[major].[minor]     |             | _dependency_                    | (12)             |                                    | \<pack version>    |
| aspnetcore-targeting-pack-[major].[minor] |             | _dependency_                    | (11)             |                                    | \<pack version>    |

> | netstandard-targeting-pack-[netst_major].[netst_minor] || _dependency_                    | (15)             |                                    | \<pack version>    |

Reviewer 2: Might be better as a list? The wrapping in the table seems pretty bad to me. Also tweaked some versions:

* `dotnet-sdk-[major]` - Latest sdk for runtime major
  * **Version:** \<sdk version>
  * **Example:** dotnet-sdk-2
  * **Dependencies:** `dotnet-sdk-[major].[latestminor]`
* `dotnet-sdk-[major].[minor]` - Latest sdk for specific runtime
  * **Version:** \<runtime version>
  * **Example:** dotnet-sdk-2.1
  * **Contains:** (3),(4)
  * **Dependencies:** `aspnetcore-runtime-[major].[minor]`, `dotnet-netcoreapp-targeting-pack[major].[minor]`, `dotnet-aspnetcore-targeting-pack[major].[minor]`, `dotnet-netstandard-targeting-pack-[netstandard_major].[netstandard_minor]`, `dotnet-apphost-pack-[major].[minor]`
* `aspnetcore-runtime-[major].[minor]` - Specific ASP.NET Core runtime
  * **Version:** \<aspnetcore runtime version>
  * **Example:** aspnetcore-runtime-2.1
  * **Contains:** (6),(7)
  * **Dependencies:** `dotnet-runtime-[major].[minor]`
* `dotnet-runtime-[major].[minor]` - Specific runtime
  * **Version:** \<runtime version>
  * **Example:** dotnet-runtime-2.1
  * **Contains:** (5)
  * **Dependencies:** `dotnet-hostfxr:<runtime version>+`
* `dotnet-hostfxr` - dependency
  * **Version:** \<runtime version>
  * **Example:** dotnet-hostfxr
  * **Contains:** (2)
  * **Dependencies:** `host:<runtime version>+`
* `dotnet-host` - dependency
  * **Version:** \<runtime version>
  * **Example:** dotnet-host
  * **Contains:** (1),(8),(9),(10)
* `dotnet-apphost-pack-[major].[minor]`
  * **Version:** \<runtime version>
  * **Contains:** (13)
* `dotnet-targeting-pack-[major].[minor]` - allows targeting a non-latest runtime
  * **Version:** \<runtime version>
  * **Contains:** (12)
* `aspnetcore-targeting-pack-[major].[minor]` - allows targeting a non-latest runtime
  * **Version:** \<aspnetcore runtime version>
  * **Contains:** (11)
* `netstandard-targeting-pack-[major].[minor]`
  * **Version:** \<netstandard version>
  * **Contains:** (15)

Some of the packages include part of the version number in their name. This allows you to install a specific version.
The rest of the version isn't included in the version name. This allows the OS package manager to update the packages (for example, automatically installing security fixes). Supported package managers are Linux specific.

The following lists the recommended packages:

> * `dotnet-sdk-[major]` - Installs latest sdk for runtime major

Reviewer 5: ```suggestion
* `dotnet-sdk-[major]` - Installs the latest sdk for runtime major
```

  * **Version:** \<sdk version>
  * **Example:** dotnet-sdk-2
  * **Dependencies:** `dotnet-sdk-[major].[latestminor]`

> * `dotnet-sdk-[major].[minor]` - Installs latest sdk for specific runtime

Reviewer 5: ```suggestion
* `dotnet-sdk-[major].[minor]` - Installs the latest sdk for specific runtime
```

  * **Version:** \<runtime version>
  * **Example:** dotnet-sdk-2.1
  * **Contains:** (3),(4)
  * **Dependencies:** `aspnetcore-runtime-[major].[minor]`, `dotnet-netcoreapp-targeting-pack[major].[minor]`, `dotnet-aspnetcore-targeting-pack[major].[minor]`, `dotnet-netstandard-targeting-pack-[netstandard_major].[netstandard_minor]`, `dotnet-apphost-pack-[major].[minor]`

> * `aspnetcore-runtime-[major].[minor]` - Installs specific ASP.NET Core runtime

Reviewer 5: ```suggestion
* `aspnetcore-runtime-[major].[minor]` - Installs a specific ASP.NET Core runtime
```

  * **Version:** \<aspnetcore runtime version>
  * **Example:** aspnetcore-runtime-2.1
  * **Contains:** (6),(7)
  * **Dependencies:** `dotnet-runtime-[major].[minor]`

> * `dotnet-runtime-deps-[major].[minor]` - Installs dependencies for running self-contained applications

Reviewer 5: ```suggestion
* `dotnet-runtime-deps-[major].[minor]` - Installs the dependencies for running self-contained applications
```

> * `dotnet-runtime-deps-[major].[minor]` - Installs the dependencies for running self-contained applications

Reviewer: I have  mixed feelings about this. This package was initially introduced so all other dotnet packages would depend on this one. And only `dotnet-runtime-deps*` would have OS-specific dependencies.

Is the description here that the goal is to "Installs the dependencies for running self-contained applications" accurate?

Author: It's not the description of the package. In the table, this was the "Use case" why someone might want to install this.

Reviewer 2: This is interesting... for the Microsoft build, it serves both purposes, but I don't know whether the "use one package for multiple distros/versions" approach is used for typical Linux distro packaging. *If* someone wants to build the SDK once to patch multiple versions of a distro, they'd hit the same problem as the Microsoft build and might want a hint in this doc that only the deps package has to be unique per version.

Reviewer: With my package maintainer hat on, this package might be trickier to provide than expected. RPM, for example, automatically identifies (using ldd, I guess) all the dependencies. As a packager, I supplement it with stuff that RPM can't see (for example, libraries loaded via `dlopen(3)`). [This is what that looks like for .NET Core Runtime 2.1](https://gist.github.com/omajid/25ab3aaa4b3bf178d1a8fe5c112aba59). Having a `runtime-deps` package means moving all those dependencies to a new package. RPM the package builder has no support for that, really. I would have to disable the automated dependency detection (which is against the Fedora packaging guidelines) and then manually identify all the dependencies and write them down in a different place. And then keep that up-to-date as .NET Core evolves.

This creates more work for a packager, makes the process more error prone, and goes against Linux distribution guidelines.

Can we have `dotnet-runtime-deps-[major].[minor]` be optional? As in, someone can provide it if they really want to, but it's not required (or strongly suggested)?

Reviewer 2: It seems to me that a self-contained app developer/user would like to be able to install `dotnet-runtime-deps-x-y` on any "has .NET Core" distro with some amount of certainty... rather than install bloat (a full `dotnet-runtime-x-y`), or maintain a list of dependencies for each distro they build for. This seems like it could be important for space-constrained scenarios.

Definitely want feedback from @leecow on this point I think. Very interesting to hear how this makes things more difficult.

Author: We can mention it is optional.
I think the main value of the package is that it makes it easy to make a container image that is small by including `dotnet-runtime-deps` and a self-contained application.
It's not a use-case that is super important to us (Red Hat) afaik.

Reviewer 2: Hmm, yeah, it seems reasonable to me to mark it optional with a description of when a package maintainer would want to provide it.

Reviewer 3: Sorry for being late to the party. I agree with the idea of describing the deps package as optional along with the use case.

  * **Version:** \<runtime version>
  * **Example:** dotnet-runtime-deps-2.1
  * **Dependencies:** _distro specific dependencies_

> * `dotnet-runtime-[major].[minor]` - Installs specific runtime

Reviewer 5: ```suggestion
* `dotnet-runtime-[major].[minor]` - Installs a specific runtime
```

  * **Version:** \<runtime version>
  * **Example:** dotnet-runtime-2.1
  * **Contains:** (5)
  * **Dependencies:** `dotnet-hostfxr:<runtime version>+`, `dotnet-runtime-deps-[major].[minor]`

* `dotnet-hostfxr` - dependency
  * **Version:** \<runtime version>
  * **Example:** dotnet-hostfxr
  * **Contains:** (2)
  * **Dependencies:** `host:<runtime version>+`

* `dotnet-host` - dependency
  * **Version:** \<runtime version>
  * **Example:** dotnet-host
  * **Contains:** (1),(8),(9),(10),(16)

* `dotnet-apphost-pack-[major].[minor]` - dependency
  * **Version:** \<runtime version>
  * **Contains:** (13)

> * `dotnet-targeting-pack-[major].[minor]` - allows targeting a non-latest runtime

Reviewer 5: ```suggestion
* `dotnet-targeting-pack-[major].[minor]` - Allows targeting a non-latest runtime
```

  * **Version:** \<runtime version>
  * **Contains:** (12)

> * `aspnetcore-targeting-pack-[major].[minor]` - allows targeting a non-latest runtime

Reviewer 5: ```suggestion
* `aspnetcore-targeting-pack-[major].[minor]` - Allows targeting a non-latest runtime
```

  * **Version:** \<aspnetcore runtime version>
  * **Contains:** (11)

> * `netstandard-targeting-pack-[major].[minor]` - allows targeting a netstandard version

Reviewer 5: ```suggestion
* `netstandard-targeting-pack-[major].[minor]` - Allows targeting a netstandard version
```
